pkg/azuredisk/nodeserver.go

```diff
@@ -349,23 +357,34 @@ func (d *Driver) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfoRequest) (
 	maxDataDiskCount := d.VolumeAttachLimit
 	if maxDataDiskCount < 0 {
 		var instanceType string
-		if runtime.GOOS == "windows" && d.cloud.UseInstanceMetadata && d.cloud.Metadata != nil {
-			metadata, err := d.cloud.Metadata.GetMetadata(azcache.CacheReadTypeDefault)
-			if err == nil && metadata.Compute != nil {
-				instanceType = metadata.Compute.VMSize
-				klog.V(5).Infof("NodeGetInfo: nodeName(%s), VM Size(%s)", d.NodeID, instanceType)
-			} else {
-				klog.Warningf("get instance type(%s) failed with: %v", d.NodeID, err)
-			}
+		var err error
+		if d.getNodeInfoFromLabels {
+			_, instanceType, err = getNodeInfoFromLabels(ctx, d.NodeID, d.cloud.KubeClient)
 		} else {
-			instances, ok := d.cloud.Instances()
-			if !ok {
-				return nil, status.Error(codes.Internal, "failed to get instances from cloud provider")
+			if runtime.GOOS == "windows" && d.cloud.UseInstanceMetadata && d.cloud.Metadata != nil {
+				metadata, err := d.cloud.Metadata.GetMetadata(azcache.CacheReadTypeDefault)
+				if err == nil && metadata.Compute != nil {
+					instanceType = metadata.Compute.VMSize
+					klog.V(5).Infof("NodeGetInfo: nodeName(%s), VM Size(%s)", d.NodeID, instanceType)
+				}
+			} else {
+				instances, ok := d.cloud.Instances()
+				if !ok {
+					klog.Warningf("failed to get instances from cloud provider")
+				} else {
+					instanceType, err = instances.InstanceType(ctx, types.NodeName(d.NodeID))
+				}
 			}
-			var err error
-			if instanceType, err = instances.InstanceType(ctx, types.NodeName(d.NodeID)); err != nil {
+			if err != nil {
 				klog.Warningf("get instance type(%s) failed with: %v", d.NodeID, err)
 			}
+			if instanceType == "" {
```

Comment:
this is interesting, in which scenarios would this if block execute? I guess if `d.getNodeInfoFromLabels` is false and somehow the other instance metadata requests timed out right?

Reply:
yes, when instance metadata requests timed out, fall back to get instance type from node labels